src/wp-includes/load.php

```diff
@@ -239,6 +239,73 @@ function wp_get_environment_type() {
 	return $current_env;
 }
 
+/**
+ * Retrieves the current runtime environment.
+ *
+ * The value can be set via the `WP_RUNTIME_ENVIRONMENT` global system variable,
+ * or a constant of the same name.
+ *
+ * The only value currently supported is 'virtualbox'. If not set, the value
+ * defaults to an empty string.
+ *
+ * @since 6.0.0
+ *
+ * @return string|false The current runtime environment type. false on failure.
+ */
+function wp_get_runtime_environment() {
+	static $current_runtime_env = '';
+	static $runtime_constant    = '';
+	static $runtime_getenv      = '';
+
+	if ( ! defined( 'WP_RUN_CORE_TESTS' ) && $current_runtime_env ) {
+		return $current_runtime_env;
+	}
+
+	$wp_runtime_environments = array( 'virtualbox' );
+
+	// Fetch the runtime environment from a constant.
+	if ( defined( 'WP_RUNTIME_ENVIRONMENT' ) ) {
+		$runtime_constant = WP_RUNTIME_ENVIRONMENT;
+	}
+
+	// Fetch the runtime environment from global system variable.
+	if ( function_exists( 'getenv' ) ) {
+		// Check if the runtime environment variable has been set, if `getenv` is available on the system.
+		$has_runtime_env = getenv( 'WP_RUNTIME_ENVIRONMENT' );
+		if ( false !== $has_runtime_env ) {
+			$runtime_getenv = $has_runtime_env;
+		}
+	}
+
+	// If WP_RUNTIME_ENVIRONMENT constant and environmental variable are different.
+	if ( ( ! empty( $runtime_constant ) && ! empty( $runtime_getenv ) )
+		&& $runtime_constant !== $runtime_getenv
+	) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			__( 'The WP_RUNTIME_ENVIRONMENT environment variable and constant do not match. Fallbacks will be used, which may negatively impact performance.' ),
+			'6.0.0'
+		);
+
+		return false;
+	}
+
+	// If set to something in $wp_runtime_environments use it, default is ''.
+	// Preference to a set constant.
+	switch ( true ) {
+		case in_array( $runtime_constant, $wp_runtime_environments, true ):
+			$current_runtime_env = $runtime_constant;
+			break;
+		case in_array( $runtime_getenv, $wp_runtime_environments, true ):
+			$current_runtime_env = $runtime_getenv;
+			break;
```

Comment:
I think we should probably change the order of these cases.

Scenarios:
- The host sets the environment variable, the dev/user sets the constant - Shouldn't we trust the host to know the system better than the dev/user?
- The dev/user sets the environment variable and the host sets the constant - This seems like an unlikely scenario when the dev/user goes to the system level and the host goes to the application level.

If the environment variable has priority, that means you could switch environments without needing to touch any files in the WP installation, etc., which acknowledges the reality of a move: `The system has changed, not the installation`

If the system tells us about itself I think we should trust it rather than trusting a line of code that was brought in from another system.

Reply:
Seems to make sense. 

@hellofromtonya @peterwilsoncc what do you think?

Reply:
> If the system tells us about itself I think we should trust it rather than trusting a line of code that was brought in from another system.

That's a good point. However, giving precedence to the constant is more consistent with `wp_get_environment_type()`. Does it makes sense for these two functions to be consistent in terms of precedence? Maybe we could change `wp_get_environment_type()` instead to also give precedence to the environment variable? Curious to see what others think.

Reply:
I appreciate Colin’s logic. I think a host is much more likely to use an environment variable than a constant. We can probably simply this function if we have a logical precedence order. 
It may not be necessary to account for potentially conflicting values.

Reply:
I'm quite worried that we're discussing this and some other architecture issues on this ticket. 

When the code was reverted in 5.9, we (the folks working on this ticket) discussed the need to get this in `early-actually-early` and with the beta release two weeks away I don't think we're hitting that goal.

I think @hellofromtonya raises some good points above about how wise it is to include this variable architectually. Once it is in, the project is stuck with it so it needs a lot of logic checking.

When `WP_ENVIRONMENT_TYPE` was introduced, it came with multiple areas in which the project could make use of it (health check, application passwords, debug defaults all quickly took advantage of it). However, even with a dedicated discussion, the variable was renamed and the original version `WP_ENVIRONMENT_TYPES` deprecated.

Reply:
> * The host sets the environment variable, the dev/user sets the constant - Shouldn't we trust the host to know the system better than the dev/user?

I was about to write `I'd go the other way and say that we should trust the site admin`, but thought again and realized that maybe the hosting company migrates them from one platform to another and sets the environment variable accordingly (and the site admin might not know whether the new platform is VB or not, but the host certainly would).  And in that case, @costdev is right...trust the host.

> Does it makes sense for these two functions to be consistent in terms of precedence? Maybe we could change wp_get_environment_type() instead to also give precedence to the environment variable? Curious to see what others think.

At first glance it seems like it would be nice if they were consistent, and I can see how it could cause confusion if they weren't.  However, since they are such different "kinds" of things (i.e., a `staging` site is still a `staging` site even if it's been migrated to a new VPS), I think it's OK if the precedence is different.  Of course, would be **really** important to document the inconsistency (including the rationale(s) for the difference).

Reply:
@peterwilsoncc VirtualBox is broken. It doesn’t handle `rename()` correctly all the time. 

I don’t think scrapping the whole rollback and more efficient use of `rename()` in `move_dir()` is a good. We have been working on this for a long time and have solutions to all the issues.

Reply:
I agree with @pbiron - precedence should not be consistent.

We need to remember who will be using these environment variables and constants.

`wp_get_environment_type()` determines whether it's local, development, staging, production. It's used primarily by the developer, but may also be used by the host if they offer a staging site facility. Not all hosts give developers access to set environment variables, and the bottom line with this function is that the developer should have the final say on the current state of the site. The environment variable is a convenience.

`wp_get_runtime_environment()` determines what we're being told about the environment. The constant is a convenience.

Host: "It's VirtualBox."
Developer: "I was using WAMP locally."

Bottom line for this function: Trust the host.

Reply:
I can appreciate that we're looking for a cleaner solution, are concerned about bloated code, inconsistent precedence, and such.

- This is a major bug.
- It's in the filesystem.
- The bug means the wrong files are deleted.
- The deletion of these files leads to errors, and possible repair costs for the site owner.
- Without this environment variable, the same bug will happen on the next update.

We've evaluated so many options that no one has any ideas for alternatives that haven't already been investigated.

IMO:
- If we find a cleaner solution, we can implement it at any time. This feature was moved to 6.0 because this issue wasn't solved. It's now solved using the only way that's worked.
- If we want to reduce bloat, remove the constant and add it if someone requests it.
- The rollback feature, and this environment variable, are being introduced for site safety. Those who have worked on this ticket are also working hard to improve all aspects of the upgrader and filesystem areas of Core. An extra environment variable is worth it.
- We can deprecate, or ignore, the environment variable at any point in the future, but let's remember: Trust the host. This is not the only issue VirtualBox has ever encountered, and it will not be the last. Being able to behave differently in a VirtualBox system is critical to Core's file operations and site safety - that much has been made clear by this bug.

Reply:
@afragen I'm not suggesting scrapping the whole of rollback. I am saying that making architectural decisions on the ticket this late in the release cycle is problematic.

I want to get this in but not at the expense of getting it right.

Reply:
Peter, this is the only solution we’ve found and believe me we looked long and hard to fix this.

Reply:
Using a constant to solve this issue was not a late decision. Adding a function to get the results of the constant similar to `wp_get_environment_type()` was and ask by others

Reply:
> I'm quite worried that we're discussing this and some other architecture issues on this ticket.

Just to clarify, I wasn't going to suggest that reconsidering the precedence in `wp_get_environment_type()` should be a part of this PR, or that the decision should be made here, I was just asking whether that makes sense to raise as a separate ticket. But it looks like the consensus is that these functions have a different purpose and don't necessarily have to be consistent, as long as the behavior is clearly documented.

Reply:
If the `Filesystem` component itself is the problem with VirtualBox, then I'd suggest moving the fix for it to a separate ticket and PR.

Why? To centralize the effort, context, and history to identify and resolve its bug or lack of compatibility. 

IMO a feature should not be "fixing" bugs, but instead be introducing new functionality and/or flows.

Separating the VB fix can also benefit the rollback feature as it'll have a clearer path for testing, consensus, and adoption while maintaining its benefits and impact separate from bug fixes.

Reply:
@SergeyBiryukov This is one of several discussions with significant implicatations.

For example, when the issue of using a constant was [first raised in Slack](https://wordpress.slack.com/archives/CULBN711P/p1638919125186800), I raised concerns about doing so immediately. Here we -- those of us working on this ticket -- discussing exactly how it should work several months later.

I also feel other concerns @hellofromtonya and I have raised are not been addressed.

Reply:
No longer using constant. Using `is_virtualbox()` to detect VB